holopy/scattering/tests/test_lens.py

```diff
@@ -0,0 +1,213 @@
+                                       scatterer, medium_wavevec, medium_index)
+        sinphi = np.sin(phi)
+        cosphi = np.cos(phi)
+        mielens_calculator = _setup_mielens_calculator(scatterer, medium_wavevec,
+                                                       medium_index)
+
+        s_perp = mielens_calculator._scat_perp_values.ravel()
+        s_prll = mielens_calculator._scat_prll_values.ravel()
+
+        S11, theta_prll = _get_smatrix_theta_near_phi_is_zero(
+                                                  s_matrix_new[:, 0, 0],
+                                                  cosphi,
+                                                  phi,
+                                                  theta)
+        S22, theta_perp = _get_smatrix_theta_near_phi_is_pi_over_2(
+                                                  s_matrix_new[:, 1, 1],
+                                                  sinphi,
+                                                  phi,
+                                                  theta)
+
+        s_perp_new = np.interp(mielens_calculator._theta_pts, theta_perp, S22)
+        s_prll_new = np.interp(mielens_calculator._theta_pts, theta_prll, S11)
+
+        assert_allclose(s_perp_new, s_perp, rtol=5e-3)
+        assert_allclose(s_prll_new, s_prll, rtol=5e-3)
+
+    def test_raw_fields_similar_mielens(self):
+        detector = test_common.xschema
+        scatterer = test_common.sphere
+        medium_wavevec = 2 * np.pi / test_common.wavelen
+        medium_index = test_common.index
+        illum_polarization = detector.illum_polarization
+
+        theory_old = MieLens(lens_angle=LENS_ANGLE)
+        pos_old = theory_old._transform_to_desired_coordinates(
+                                detector, scatterer.center, wavevec=medium_wavevec)
+
+        theory_new = LENSMIE
+        pos_new = theory_new._transform_to_desired_coordinates(
+                                detector, scatterer.center, wavevec=medium_wavevec)
+
+        f0x, f0y, f0z = theory_old._raw_fields(pos_old, scatterer, medium_wavevec, medium_index,
+                        illum_polarization)
+        fx, fy, fz = theory_new._raw_fields(pos_new, scatterer, medium_wavevec, medium_index,
+                        illum_polarization)
+        assert_allclose(f0x, fx, atol=2e-3)
+        assert_allclose(f0y, fy, atol=2e-3)
+        assert_allclose(f0z, fz, atol=2e-3)
+
+    def test_lens_plus_mie_fields_same_as_mielens(self):
+        detector = test_common.xschema
+        scatterer = test_common.sphere
+        medium_wavevec = 2 * np.pi / test_common.wavelen
+        medium_index = test_common.index
+        illum_polarization = test_common.xpolarization
+
+        theory_old = MieLens(lens_angle=LENS_ANGLE)
+        theory_new = LENSMIE
+
+        fields_old = theory_old.calculate_scattered_field(scatterer, detector)
+        fields_new = theory_new.calculate_scattered_field(scatterer, detector)
+
```

Comment:
I think the above tests are good, and you should keep them. 

But I think we also need a test that MieLens() and Lens(MIe) are the same at much tighter tolerances, say at 1e-12 levels. This might require increasing the number of quadrature points by a lot for both, so I don't think this is a test that we want to run over a full 64x64 hologram; we should instead test it over a few points (say 5 randomly chosen points at some random z).

The reason that I think this is important is that there are some "small signal" physical effects, which either Lens(Mie) or MieLens could be getting backwards. For instance, there is a cos(2phi) term in the hologram from a sphere, but it is only noticeable at about the 1e-3 level (depending on parameters). So it's possible for one of the two theories to have a sign error and for these tests to still pass.

Reply:
I have resolved everything except this comment. It’s not so simple as adjusting test parameters to be 5 pixels and “enough” quadrature points since the numpy guassian quadrature algotithm takes time too.

1000x1000 hologram + 100x100 quadrature takes less way less time than 100x100 hologram + 1000x1000 quadrature.

Reply:
Hmm... how much does the deviation between Lens(Mie) and MieLens() improve if you do (say) 200x200 quadrature?

I think it's important to have some sort of test that at least checks that the deviation decreases between the two equivalent theories as the number of quad pts increases. We don't have to go down to 1e-13 tolerances, but something a little tighter than 5e-3 -- something like 1e-5 or 1e-6 should be fine.

If you want, I can take a look at it. Just let me know; I should have some time tomorrow.

Reply:
Going from 100x100 pixels to 200x200 gives less than x2 of precision but takes more than x2 time. Try it out tomorrow and you can see what I mean.

Reply:
OK... so I tested whether MieLens() and Lens(Mie()) go to the same values as the number of quadrature points gets large, and the results are weird. 

Basically, both Lens(Mie) and MieLens() converge very rapidly at small distances as the number of quadrature points is increased. However, they do not converge to the same values. When the quadrature points of both are large, the fields differ by ~6e-5 (cf. the new test in e68f0ea ). This is a weird number. 6e-5 is smaller than any physics, but much bigger than numerics should be. And while the difference in the fields is not random, there is no obvious pattern in the differences either (e.g. it's not cos(2phi) symmetry). 

So... I don't know what to do. Right now I'm thinking leave the failing test in there, skipped, merge this in, and raise an issue that there is a subtle problem. My hunch is that the difference is in the computation of the Mie scattering coefficients....

Thoughs @ralex0  and @barkls ?

Reply:
I worry about leaving in skipped failing tests because that seems a bit like pushing the problem onto some future HoloPy developer who probably doesn't care too much about that specific issue. I would prefer to have a passing test (with more forgiving tolerance), and raise an issue that it should be looked into. That way there is still a record of the problem and that it needs to be fixed, but we are running the test in the meantime.